<!doctype html>
<html>
<head>
<meta charset='UTF-8'><meta name='viewport' content='width=device-width initial-scale=1'>
<title>Report</title><link href='https://fonts.loli.net/css?family=Open+Sans:400italic,700italic,700,400&subset=latin,latin-ext' rel='stylesheet' type='text/css' /><style type='text/css'>html {overflow-x: initial !important;}:root { --bg-color:#ffffff; --text-color:#333333; --select-text-bg-color:#B5D6FC; --select-text-font-color:auto; --monospace:"Lucida Console",Consolas,"Courier",monospace; }
html { font-size: 14px; background-color: var(--bg-color); color: var(--text-color); font-family: "Helvetica Neue", Helvetica, Arial, sans-serif; -webkit-font-smoothing: antialiased; }
body { margin: 0px; padding: 0px; height: auto; bottom: 0px; top: 0px; left: 0px; right: 0px; font-size: 1rem; line-height: 1.42857; overflow-x: hidden; background: inherit; tab-size: 4; }
iframe { margin: auto; }
a.url { word-break: break-all; }
a:active, a:hover { outline: 0px; }
.in-text-selection, ::selection { text-shadow: none; background: var(--select-text-bg-color); color: var(--select-text-font-color); }
#write { margin: 0px auto; height: auto; width: inherit; word-break: normal; overflow-wrap: break-word; position: relative; white-space: normal; overflow-x: visible; padding-top: 40px; }
#write.first-line-indent p { text-indent: 2em; }
#write.first-line-indent li p, #write.first-line-indent p * { text-indent: 0px; }
#write.first-line-indent li { margin-left: 2em; }
.for-image #write { padding-left: 8px; padding-right: 8px; }
body.typora-export { padding-left: 30px; padding-right: 30px; }
.typora-export .footnote-line, .typora-export li, .typora-export p { white-space: pre-wrap; }
@media screen and (max-width: 500px) {
  body.typora-export { padding-left: 0px; padding-right: 0px; }
  #write { padding-left: 20px; padding-right: 20px; }
  .CodeMirror-sizer { margin-left: 0px !important; }
  .CodeMirror-gutters { display: none !important; }
}
#write li > figure:last-child { margin-bottom: 0.5rem; }
#write ol, #write ul { position: relative; }
img { max-width: 100%; vertical-align: middle; }
button, input, select, textarea { color: inherit; font: inherit; }
input[type="checkbox"], input[type="radio"] { line-height: normal; padding: 0px; }
*, ::after, ::before { box-sizing: border-box; }
#write h1, #write h2, #write h3, #write h4, #write h5, #write h6, #write p, #write pre { width: inherit; }
#write h1, #write h2, #write h3, #write h4, #write h5, #write h6, #write p { position: relative; }
p { line-height: inherit; }
h1, h2, h3, h4, h5, h6 { break-after: avoid-page; break-inside: avoid; orphans: 2; }
p { orphans: 4; }
h1 { font-size: 2rem; }
h2 { font-size: 1.8rem; }
h3 { font-size: 1.6rem; }
h4 { font-size: 1.4rem; }
h5 { font-size: 1.2rem; }
h6 { font-size: 1rem; }
.md-math-block, .md-rawblock, h1, h2, h3, h4, h5, h6, p { margin-top: 1rem; margin-bottom: 1rem; }
.hidden { display: none; }
.md-blockmeta { color: rgb(204, 204, 204); font-weight: 700; font-style: italic; }
a { cursor: pointer; }
sup.md-footnote { padding: 2px 4px; background-color: rgba(238, 238, 238, 0.7); color: rgb(85, 85, 85); border-radius: 4px; cursor: pointer; }
sup.md-footnote a, sup.md-footnote a:hover { color: inherit; text-transform: inherit; text-decoration: inherit; }
#write input[type="checkbox"] { cursor: pointer; width: inherit; height: inherit; }
figure { overflow-x: auto; margin: 1.2em 0px; max-width: calc(100% + 16px); padding: 0px; }
figure > table { margin: 0px !important; }
tr { break-inside: avoid; break-after: auto; }
thead { display: table-header-group; }
table { border-collapse: collapse; border-spacing: 0px; width: 100%; overflow: auto; break-inside: auto; text-align: left; }
table.md-table td { min-width: 32px; }
.CodeMirror-gutters { border-right: 0px; background-color: inherit; }
.CodeMirror-linenumber { user-select: none; }
.CodeMirror { text-align: left; }
.CodeMirror-placeholder { opacity: 0.3; }
.CodeMirror pre { padding: 0px 4px; }
.CodeMirror-lines { padding: 0px; }
div.hr:focus { cursor: none; }
#write pre { white-space: pre-wrap; }
#write.fences-no-line-wrapping pre { white-space: pre; }
#write pre.ty-contain-cm { white-space: normal; }
.CodeMirror-gutters { margin-right: 4px; }
.md-fences { font-size: 0.9rem; display: block; break-inside: avoid; text-align: left; overflow: visible; white-space: pre; background: inherit; position: relative !important; }
.md-diagram-panel { width: 100%; margin-top: 10px; text-align: center; padding-top: 0px; padding-bottom: 8px; overflow-x: auto; }
#write .md-fences.mock-cm { white-space: pre-wrap; }
.md-fences.md-fences-with-lineno { padding-left: 0px; }
#write.fences-no-line-wrapping .md-fences.mock-cm { white-space: pre; overflow-x: auto; }
.md-fences.mock-cm.md-fences-with-lineno { padding-left: 8px; }
.CodeMirror-line, twitterwidget { break-inside: avoid; }
.footnotes { opacity: 0.8; font-size: 0.9rem; margin-top: 1em; margin-bottom: 1em; }
.footnotes + .footnotes { margin-top: 0px; }
.md-reset { margin: 0px; padding: 0px; border: 0px; outline: 0px; vertical-align: top; background: 0px 0px; text-decoration: none; text-shadow: none; float: none; position: static; width: auto; height: auto; white-space: nowrap; cursor: inherit; -webkit-tap-highlight-color: transparent; line-height: normal; font-weight: 400; text-align: left; box-sizing: content-box; direction: ltr; }
li div { padding-top: 0px; }
blockquote { margin: 1rem 0px; }
li .mathjax-block, li p { margin: 0.5rem 0px; }
li { margin: 0px; position: relative; }
blockquote > :last-child { margin-bottom: 0px; }
blockquote > :first-child, li > :first-child { margin-top: 0px; }
.footnotes-area { color: rgb(136, 136, 136); margin-top: 0.714rem; padding-bottom: 0.143rem; white-space: normal; }
#write .footnote-line { white-space: pre-wrap; }
@media print {
  body, html { border: 1px solid transparent; height: 99%; break-after: avoid; break-before: avoid; }
  #write { margin-top: 0px; padding-top: 0px; border-color: transparent !important; }
  .typora-export * { -webkit-print-color-adjust: exact; }
  html.blink-to-pdf { font-size: 13px; }
  .typora-export #write { padding-left: 32px; padding-right: 32px; padding-bottom: 0px; break-after: avoid; }
  .typora-export #write::after { height: 0px; }
}
.footnote-line { margin-top: 0.714em; font-size: 0.7em; }
a img, img a { cursor: pointer; }
pre.md-meta-block { font-size: 0.8rem; min-height: 0.8rem; white-space: pre-wrap; background: rgb(204, 204, 204); display: block; overflow-x: hidden; }
p > .md-image:only-child:not(.md-img-error) img, p > img:only-child { display: block; margin: auto; }
p > .md-image:only-child { display: inline-block; width: 100%; }
#write .MathJax_Display { margin: 0.8em 0px 0px; }
.md-math-block { width: 100%; }
.md-math-block:not(:empty)::after { display: none; }
[contenteditable="true"]:active, [contenteditable="true"]:focus, [contenteditable="false"]:active, [contenteditable="false"]:focus { outline: 0px; box-shadow: none; }
.md-task-list-item { position: relative; list-style-type: none; }
.task-list-item.md-task-list-item { padding-left: 0px; }
.md-task-list-item > input { position: absolute; top: 0px; left: 0px; margin-left: -1.2em; margin-top: calc(1em - 10px); border: none; }
.math { font-size: 1rem; }
.md-toc { min-height: 3.58rem; position: relative; font-size: 0.9rem; border-radius: 10px; }
.md-toc-content { position: relative; margin-left: 0px; }
.md-toc-content::after, .md-toc::after { display: none; }
.md-toc-item { display: block; color: rgb(65, 131, 196); }
.md-toc-item a { text-decoration: none; }
.md-toc-inner:hover { text-decoration: underline; }
.md-toc-inner { display: inline-block; cursor: pointer; }
.md-toc-h1 .md-toc-inner { margin-left: 0px; font-weight: 700; }
.md-toc-h2 .md-toc-inner { margin-left: 2em; }
.md-toc-h3 .md-toc-inner { margin-left: 4em; }
.md-toc-h4 .md-toc-inner { margin-left: 6em; }
.md-toc-h5 .md-toc-inner { margin-left: 8em; }
.md-toc-h6 .md-toc-inner { margin-left: 10em; }
@media screen and (max-width: 48em) {
  .md-toc-h3 .md-toc-inner { margin-left: 3.5em; }
  .md-toc-h4 .md-toc-inner { margin-left: 5em; }
  .md-toc-h5 .md-toc-inner { margin-left: 6.5em; }
  .md-toc-h6 .md-toc-inner { margin-left: 8em; }
}
a.md-toc-inner { font-size: inherit; font-style: inherit; font-weight: inherit; line-height: inherit; }
.footnote-line a:not(.reversefootnote) { color: inherit; }
.md-attr { display: none; }
.md-fn-count::after { content: "."; }
code, pre, samp, tt { font-family: var(--monospace); }
kbd { margin: 0px 0.1em; padding: 0.1em 0.6em; font-size: 0.8em; color: rgb(36, 39, 41); background: rgb(255, 255, 255); border: 1px solid rgb(173, 179, 185); border-radius: 3px; box-shadow: rgba(12, 13, 14, 0.2) 0px 1px 0px, rgb(255, 255, 255) 0px 0px 0px 2px inset; white-space: nowrap; vertical-align: middle; }
.md-comment { color: rgb(162, 127, 3); opacity: 0.8; font-family: var(--monospace); }
code { text-align: left; vertical-align: initial; }
a.md-print-anchor { white-space: pre !important; border-width: initial !important; border-style: none !important; border-color: initial !important; display: inline-block !important; position: absolute !important; width: 1px !important; right: 0px !important; outline: 0px !important; background: 0px 0px !important; text-decoration: initial !important; text-shadow: initial !important; }
.md-inline-math .MathJax_SVG .noError { display: none !important; }
.html-for-mac .inline-math-svg .MathJax_SVG { vertical-align: 0.2px; }
.md-math-block .MathJax_SVG_Display { text-align: center; margin: 0px; position: relative; text-indent: 0px; max-width: none; max-height: none; min-height: 0px; min-width: 100%; width: auto; overflow-y: hidden; display: block !important; }
.MathJax_SVG_Display, .md-inline-math .MathJax_SVG_Display { width: auto; margin: inherit; display: inline-block !important; }
.MathJax_SVG .MJX-monospace { font-family: var(--monospace); }
.MathJax_SVG .MJX-sans-serif { font-family: sans-serif; }
.MathJax_SVG { display: inline; font-style: normal; font-weight: 400; line-height: normal; zoom: 90%; text-indent: 0px; text-align: left; text-transform: none; letter-spacing: normal; word-spacing: normal; overflow-wrap: normal; white-space: nowrap; float: none; direction: ltr; max-width: none; max-height: none; min-width: 0px; min-height: 0px; border: 0px; padding: 0px; margin: 0px; }
.MathJax_SVG * { transition: none 0s ease 0s; }
.MathJax_SVG_Display svg { vertical-align: middle !important; margin-bottom: 0px !important; margin-top: 0px !important; }
.os-windows.monocolor-emoji .md-emoji { font-family: "Segoe UI Symbol", sans-serif; }
.md-diagram-panel > svg { max-width: 100%; }
[lang="flow"] svg, [lang="mermaid"] svg { max-width: 100%; height: auto; }
[lang="mermaid"] .node text { font-size: 1rem; }
table tr th { border-bottom: 0px; }
video { max-width: 100%; display: block; margin: 0px auto; }
iframe { max-width: 100%; width: 100%; border: none; }
.highlight td, .highlight tr { border: 0px; }
svg[id^="mermaidChart"] { line-height: 1em; }
mark { background: rgb(255, 255, 0); color: rgb(0, 0, 0); }
.md-html-inline .md-plain, .md-html-inline strong, mark .md-inline-math, mark strong { color: inherit; }
mark .md-meta { color: rgb(0, 0, 0); opacity: 0.3 !important; }


.CodeMirror { height: auto; }
.CodeMirror.cm-s-inner { background: inherit; }
.CodeMirror-scroll { overflow: auto hidden; z-index: 3; }
.CodeMirror-gutter-filler, .CodeMirror-scrollbar-filler { background-color: rgb(255, 255, 255); }
.CodeMirror-gutters { border-right: 1px solid rgb(221, 221, 221); background: inherit; white-space: nowrap; }
.CodeMirror-linenumber { padding: 0px 3px 0px 5px; text-align: right; color: rgb(153, 153, 153); }
.cm-s-inner .cm-keyword { color: rgb(119, 0, 136); }
.cm-s-inner .cm-atom, .cm-s-inner.cm-atom { color: rgb(34, 17, 153); }
.cm-s-inner .cm-number { color: rgb(17, 102, 68); }
.cm-s-inner .cm-def { color: rgb(0, 0, 255); }
.cm-s-inner .cm-variable { color: rgb(0, 0, 0); }
.cm-s-inner .cm-variable-2 { color: rgb(0, 85, 170); }
.cm-s-inner .cm-variable-3 { color: rgb(0, 136, 85); }
.cm-s-inner .cm-string { color: rgb(170, 17, 17); }
.cm-s-inner .cm-property { color: rgb(0, 0, 0); }
.cm-s-inner .cm-operator { color: rgb(152, 26, 26); }
.cm-s-inner .cm-comment, .cm-s-inner.cm-comment { color: rgb(170, 85, 0); }
.cm-s-inner .cm-string-2 { color: rgb(255, 85, 0); }
.cm-s-inner .cm-meta { color: rgb(85, 85, 85); }
.cm-s-inner .cm-qualifier { color: rgb(85, 85, 85); }
.cm-s-inner .cm-builtin { color: rgb(51, 0, 170); }
.cm-s-inner .cm-bracket { color: rgb(153, 153, 119); }
.cm-s-inner .cm-tag { color: rgb(17, 119, 0); }
.cm-s-inner .cm-attribute { color: rgb(0, 0, 204); }
.cm-s-inner .cm-header, .cm-s-inner.cm-header { color: rgb(0, 0, 255); }
.cm-s-inner .cm-quote, .cm-s-inner.cm-quote { color: rgb(0, 153, 0); }
.cm-s-inner .cm-hr, .cm-s-inner.cm-hr { color: rgb(153, 153, 153); }
.cm-s-inner .cm-link, .cm-s-inner.cm-link { color: rgb(0, 0, 204); }
.cm-negative { color: rgb(221, 68, 68); }
.cm-positive { color: rgb(34, 153, 34); }
.cm-header, .cm-strong { font-weight: 700; }
.cm-del { text-decoration: line-through; }
.cm-em { font-style: italic; }
.cm-link { text-decoration: underline; }
.cm-error { color: red; }
.cm-invalidchar { color: red; }
.cm-constant { color: rgb(38, 139, 210); }
.cm-defined { color: rgb(181, 137, 0); }
div.CodeMirror span.CodeMirror-matchingbracket { color: rgb(0, 255, 0); }
div.CodeMirror span.CodeMirror-nonmatchingbracket { color: rgb(255, 34, 34); }
.cm-s-inner .CodeMirror-activeline-background { background: inherit; }
.CodeMirror { position: relative; overflow: hidden; }
.CodeMirror-scroll { height: 100%; outline: 0px; position: relative; box-sizing: content-box; background: inherit; }
.CodeMirror-sizer { position: relative; }
.CodeMirror-gutter-filler, .CodeMirror-hscrollbar, .CodeMirror-scrollbar-filler, .CodeMirror-vscrollbar { position: absolute; z-index: 6; display: none; }
.CodeMirror-vscrollbar { right: 0px; top: 0px; overflow: hidden; }
.CodeMirror-hscrollbar { bottom: 0px; left: 0px; overflow: hidden; }
.CodeMirror-scrollbar-filler { right: 0px; bottom: 0px; }
.CodeMirror-gutter-filler { left: 0px; bottom: 0px; }
.CodeMirror-gutters { position: absolute; left: 0px; top: 0px; padding-bottom: 30px; z-index: 3; }
.CodeMirror-gutter { white-space: normal; height: 100%; box-sizing: content-box; padding-bottom: 30px; margin-bottom: -32px; display: inline-block; }
.CodeMirror-gutter-wrapper { position: absolute; z-index: 4; background: 0px 0px !important; border: none !important; }
.CodeMirror-gutter-background { position: absolute; top: 0px; bottom: 0px; z-index: 4; }
.CodeMirror-gutter-elt { position: absolute; cursor: default; z-index: 4; }
.CodeMirror-lines { cursor: text; }
.CodeMirror pre { border-radius: 0px; border-width: 0px; background: 0px 0px; font-family: inherit; font-size: inherit; margin: 0px; white-space: pre; overflow-wrap: normal; color: inherit; z-index: 2; position: relative; overflow: visible; }
.CodeMirror-wrap pre { overflow-wrap: break-word; white-space: pre-wrap; word-break: normal; }
.CodeMirror-code pre { border-right: 30px solid transparent; width: fit-content; }
.CodeMirror-wrap .CodeMirror-code pre { border-right: none; width: auto; }
.CodeMirror-linebackground { position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px; z-index: 0; }
.CodeMirror-linewidget { position: relative; z-index: 2; overflow: auto; }
.CodeMirror-wrap .CodeMirror-scroll { overflow-x: hidden; }
.CodeMirror-measure { position: absolute; width: 100%; height: 0px; overflow: hidden; visibility: hidden; }
.CodeMirror-measure pre { position: static; }
.CodeMirror div.CodeMirror-cursor { position: absolute; visibility: hidden; border-right: none; width: 0px; }
.CodeMirror div.CodeMirror-cursor { visibility: hidden; }
.CodeMirror-focused div.CodeMirror-cursor { visibility: inherit; }
.cm-searching { background: rgba(255, 255, 0, 0.4); }
@media print {
  .CodeMirror div.CodeMirror-cursor { visibility: hidden; }
}


:root {
    --side-bar-bg-color: #fafafa;
    --control-text-color: #777;
}

@include-when-export url(https://fonts.loli.net/css?family=Open+Sans:400italic,700italic,700,400&subset=latin,latin-ext);

html {
    font-size: 16px;
}

body {
    font-family: "Open Sans","Clear Sans","Helvetica Neue",Helvetica,Arial,sans-serif;
    color: rgb(51, 51, 51);
    line-height: 1.6;
}

#write {
    max-width: 860px;
  	margin: 0 auto;
  	padding: 30px;
    padding-bottom: 100px;
}
#write > ul:first-child,
#write > ol:first-child{
    margin-top: 30px;
}

a {
    color: #4183C4;
}
h1,
h2,
h3,
h4,
h5,
h6 {
    position: relative;
    margin-top: 1rem;
    margin-bottom: 1rem;
    font-weight: bold;
    line-height: 1.4;
    cursor: text;
}
h1:hover a.anchor,
h2:hover a.anchor,
h3:hover a.anchor,
h4:hover a.anchor,
h5:hover a.anchor,
h6:hover a.anchor {
    text-decoration: none;
}
h1 tt,
h1 code {
    font-size: inherit;
}
h2 tt,
h2 code {
    font-size: inherit;
}
h3 tt,
h3 code {
    font-size: inherit;
}
h4 tt,
h4 code {
    font-size: inherit;
}
h5 tt,
h5 code {
    font-size: inherit;
}
h6 tt,
h6 code {
    font-size: inherit;
}
h1 {
    padding-bottom: .3em;
    font-size: 2.25em;
    line-height: 1.2;
    border-bottom: 1px solid #eee;
}
h2 {
   padding-bottom: .3em;
    font-size: 1.75em;
    line-height: 1.225;
    border-bottom: 1px solid #eee;
}
h3 {
    font-size: 1.5em;
    line-height: 1.43;
}
h4 {
    font-size: 1.25em;
}
h5 {
    font-size: 1em;
}
h6 {
   font-size: 1em;
    color: #777;
}
p,
blockquote,
ul,
ol,
dl,
table{
    margin: 0.8em 0;
}
li>ol,
li>ul {
    margin: 0 0;
}
hr {
    height: 2px;
    padding: 0;
    margin: 16px 0;
    background-color: #e7e7e7;
    border: 0 none;
    overflow: hidden;
    box-sizing: content-box;
}

li p.first {
    display: inline-block;
}
ul,
ol {
    padding-left: 30px;
}
ul:first-child,
ol:first-child {
    margin-top: 0;
}
ul:last-child,
ol:last-child {
    margin-bottom: 0;
}
blockquote {
    border-left: 4px solid #dfe2e5;
    padding: 0 15px;
    color: #777777;
}
blockquote blockquote {
    padding-right: 0;
}
table {
    padding: 0;
    word-break: initial;
}
table tr {
    border-top: 1px solid #dfe2e5;
    margin: 0;
    padding: 0;
}
table tr:nth-child(2n),
thead {
    background-color: #f8f8f8;
}
table tr th {
    font-weight: bold;
    border: 1px solid #dfe2e5;
    border-bottom: 0;
    margin: 0;
    padding: 6px 13px;
}
table tr td {
    border: 1px solid #dfe2e5;
    margin: 0;
    padding: 6px 13px;
}
table tr th:first-child,
table tr td:first-child {
    margin-top: 0;
}
table tr th:last-child,
table tr td:last-child {
    margin-bottom: 0;
}

.CodeMirror-lines {
    padding-left: 4px;
}

.code-tooltip {
    box-shadow: 0 1px 1px 0 rgba(0,28,36,.3);
    border-top: 1px solid #eef2f2;
}

.md-fences,
code,
tt {
    border: 1px solid #e7eaed;
    background-color: #f8f8f8;
    border-radius: 3px;
    padding: 0;
    padding: 2px 4px 0px 4px;
    font-size: 0.9em;
}

code {
    background-color: #f3f4f4;
    padding: 0 2px 0 2px;
}

.md-fences {
    margin-bottom: 15px;
    margin-top: 15px;
    padding-top: 8px;
    padding-bottom: 6px;
}


.md-task-list-item > input {
  margin-left: -1.3em;
}

@media print {
    html {
        font-size: 13px;
    }
    table,
    pre {
        page-break-inside: avoid;
    }
    pre {
        word-wrap: break-word;
    }
}

.md-fences {
	background-color: #f8f8f8;
}
#write pre.md-meta-block {
	padding: 1rem;
    font-size: 85%;
    line-height: 1.45;
    background-color: #f7f7f7;
    border: 0;
    border-radius: 3px;
    color: #777777;
    margin-top: 0 !important;
}

.mathjax-block>.code-tooltip {
	bottom: .375rem;
}

.md-mathjax-midline {
    background: #fafafa;
}

#write>h3.md-focus:before{
	left: -1.5625rem;
	top: .375rem;
}
#write>h4.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
#write>h5.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
#write>h6.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
.md-image>.md-meta {
    /*border: 1px solid #ddd;*/
    border-radius: 3px;
    padding: 2px 0px 0px 4px;
    font-size: 0.9em;
    color: inherit;
}

.md-tag {
    color: #a7a7a7;
    opacity: 1;
}

.md-toc { 
    margin-top:20px;
    padding-bottom:20px;
}

.sidebar-tabs {
    border-bottom: none;
}

#typora-quick-open {
    border: 1px solid #ddd;
    background-color: #f8f8f8;
}

#typora-quick-open-item {
    background-color: #FAFAFA;
    border-color: #FEFEFE #e5e5e5 #e5e5e5 #eee;
    border-style: solid;
    border-width: 1px;
}

/** focus mode */
.on-focus-mode blockquote {
    border-left-color: rgba(85, 85, 85, 0.12);
}

header, .context-menu, .megamenu-content, footer{
    font-family: "Segoe UI", "Arial", sans-serif;
}

.file-node-content:hover .file-node-icon,
.file-node-content:hover .file-node-open-state{
    visibility: visible;
}

.mac-seamless-mode #typora-sidebar {
    background-color: #fafafa;
    background-color: var(--side-bar-bg-color);
}

.md-lang {
    color: #b4654d;
}

.html-for-mac .context-menu {
    --item-hover-bg-color: #E6F0FE;
}

#md-notification .btn {
    border: 0;
}

.dropdown-menu .divider {
    border-color: #e5e5e5;
}

.ty-preferences .window-content {
    background-color: #fafafa;
}

.ty-preferences .nav-group-item.active {
    color: white;
    background: #999;
}


</style>
</head>
<body class='typora-export' >
<div  id='write'  class = 'is-node'><h1><a name="centerspan-styletext-aligncenter--colorbluereport-of-your-samba-metabarcoding-analysisspancenter" class="md-header-anchor"></a><center><span style='text-align:center ; color:blue'><span>Report of your SAMBA metabarcoding analysis</span></span></center></h1><p>&nbsp;</p><p><span style="color:black"><u><strong><span>Authors:</span></strong></u><span> Laure QUINTRIC &amp; Cyril NOËL - Bioinformatics engineers  - SeBiMER (Ifremer)</span></span><span> </span></p><p><span style="color:black"><u><strong><span>Contact:</span></strong></u><span> </span><em><span>samba-sebimer@ifremer.fr</span></em></span><span> </span></p><div class='md-toc' mdtype='toc'><p class="md-toc-content" role="list"><span role="listitem" class="md-toc-item md-toc-h1" data-ref="n0"><a class="md-toc-inner" href="#centerspan-styletext-aligncenter--colorbluereport-of-your-samba-metabarcoding-analysisspancenter">Report of your SAMBA metabarcoding analysis</a></span><span role="listitem" class="md-toc-item md-toc-h2" data-ref="n6"><a class="md-toc-inner" href="#span-stylecolorredi-report-introductionspan">I. Report introduction</a></span><span role="listitem" class="md-toc-item md-toc-h2" data-ref="n10"><a class="md-toc-inner" href="#span-stylecolorredii-description-of-the-workflowspan">II. Description of the workflow</a></span><span role="listitem" class="md-toc-item md-toc-h2" data-ref="n14"><a class="md-toc-inner" href="#span-stylecolorrediii-bioinformatic-processspan">III. Bioinformatic process</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n15"><a class="md-toc-inner" href="#iii1-data-integrity">III.1. Data integrity</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n23"><a class="md-toc-inner" href="#iii2-importing-raw-data">III.2. Importing raw data</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n37"><a class="md-toc-inner" href="#iii3-removal-of-primers">III.3. Removal of primers</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n53"><a class="md-toc-inner" href="#iii4--sequence-quality-control-and-feature-table-construction">III.4.  Sequence quality control and feature table construction</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n67"><a class="md-toc-inner" href="#iii5--asv-clustering">III.5.  ASV clustering</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n81"><a class="md-toc-inner" href="#iii6--taxonomic-assignation">III.6.  Taxonomic assignation</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n95"><a class="md-toc-inner" href="#iii7--samples-decontamination-microdecon-optional">III.7.  Samples decontamination (microDecon) [optional]</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n107"><a class="md-toc-inner" href="#iii8--functional-predictions-picrust2-optional">III.8.  Functional predictions (PICRUSt2) [optional]</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n121"><a class="md-toc-inner" href="#iii9--final-outputs">III.9.  Final outputs</a></span><span role="listitem" class="md-toc-item md-toc-h2" data-ref="n130"><a class="md-toc-inner" href="#span-stylecolorrediv-general-statistical-analysesspan">IV. General statistical analyses</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n134"><a class="md-toc-inner" href="#iv1--alpha-diversity">IV.1.  Alpha diversity</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n135"><a class="md-toc-inner" href="#iiv1a-diversity-indicesi">IV.1.a diversity indices</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n137"><a class="md-toc-inner" href="#iiv1b-rarefaction-curvei">IV.1.b rarefaction curve</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n139"><a class="md-toc-inner" href="#iiv1c-taxonomic-diversityi">IV.1.c taxonomic diversity</a></span><span role="listitem" class="md-toc-item md-toc-h3" data-ref="n147"><a class="md-toc-inner" href="#iv2--beta-diversity">IV.2.  Beta diversity</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n151"><a class="md-toc-inner" href="#iiv2a-nmdsi">IV.2.a NMDS </a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n152"><a class="md-toc-inner" href="#iv2ai-rarefied-data">IV.2.a.i rarefied data</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n155"><a class="md-toc-inner" href="#iv2aii-deseq2-normalization">IV.2.a.ii DESeq2 normalization</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n158"><a class="md-toc-inner" href="#iv2aiii-css-normalization">IV.2.a.iii CSS normalization</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n161"><a class="md-toc-inner" href="#iiv2b-pcoai">IV.2.b PCoA </a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n162"><a class="md-toc-inner" href="#iv2bi-rarefied-data">IV.2.b.i rarefied data</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n165"><a class="md-toc-inner" href="#iv2bii-deseq2-normalization">IV.2.b.ii DESeq2 normalization</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n168"><a class="md-toc-inner" href="#iv2biii-css-normalization">IV.2.b.iii CSS normalization</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n171"><a class="md-toc-inner" href="#iiv2c-hierarchical-clusteringi">IV.2.c Hierarchical clustering </a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n172"><a class="md-toc-inner" href="#iv2ci-rarefied-data">IV.2.c.i rarefied data</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n175"><a class="md-toc-inner" href="#iv2cii-deseq2-normalization">IV.2.c.ii DESeq2 normalization</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n178"><a class="md-toc-inner" href="#iv2ciii-css-normalization">IV.2.c.iii CSS normalization</a></span><span role="listitem" class="md-toc-item md-toc-h4" data-ref="n181"><a class="md-toc-inner" href="#iiv2d-explained-variancei">IV.2.d Explained variance </a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n182"><a class="md-toc-inner" href="#iv2di-rarefied-data">IV.2.d.i rarefied data</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n185"><a class="md-toc-inner" href="#iv2dii-deseq2-normalization">IV.2.d.ii DESeq2 normalization</a></span><span role="listitem" class="md-toc-item md-toc-h6" data-ref="n188"><a class="md-toc-inner" href="#iv2diii-css-normalization">IV.2.d.iii CSS normalization</a></span></p></div><h2><a name="span-stylecolorredi-report-introductionspan" class="md-header-anchor"></a><span style="color:red"><span>I. Report introduction</span></span></h2><div style="text-align: justify">
   This document is the report of your metabarcoding analysis performed using using 
   <b>the SAMBA workflow</b> created by the engineers of the Ifremer's Bioinformatics 
   service (<b>SeBiMER</b>).
   This report includes results in the form of clickable links referring to generated 
   analysis files. In order not to break them, <b>be sure not to move this document 
   independently of all the result folder</b>.
</div><p>&nbsp;</p><h2><a name="span-stylecolorredii-description-of-the-workflowspan" class="md-header-anchor"></a><span style="color:red"><span>II. Description of the workflow</span></span></h2><div style="text-align: justify">
   The <i>SAMBA workflow</i> developed by <u>Laure Quintric and Cyril Noël</u> allows 
   you to <b>computerize metabarcoding analysis </b> (16S, 18S, ITS) using the 
   <b>DADA2 package through QIIME 2</b> (Bolyen <i>et al.</i>, 2019 ; version 2019.10). 
   It also performs <b>basic diversity analyses</b> (alpha and beta) via the use of 
   <font color="green">homemade R scripts</font> using the vegan, phyloseq and ggplot2 
   R packages. 
   Specific statistics based on your metadata can also be carried out either by using, 
   in your own scripts, the <i>phyloseq</i> object created by SAMBA (advanced users) or 
   by coming back to us for additional analyses.
   This workflow is based on the use of the <b>NextFlow</b> workflow managers. All 
   analysis parameters are fully configurable via the 
   <font color="red">params.config file</font> available in the config folder of SAMBA.
</div><p>&nbsp;</p><h2><a name="span-stylecolorrediii-bioinformatic-processspan" class="md-header-anchor"></a><span style="color:red"><span>III. Bioinformatic process</span></span></h2><h3><a name="iii1-data-integrity" class="md-header-anchor"></a><span>III.1. Data integrity</span></h3><div style="text-align: justify"><span style="color:black">
   This first step allows to <b>analyze the integrity of your data</b> in order to identify 
   potential problems related to the sequencing processes. It checks:
<ul>
    <li>that each read is correctly associated with the proper sample (sequence barcode verification)</li>
    <li>that they come from a single sequencer</li>
    <li>the efficiency of forward and reverse PCR amplification</li>
</ul>
   <b>This step is carried out by a <font color="green">homemade script</font></b>
</span></div><p>&nbsp;</p><div style="background-color:yellow"><center><b>RESULTS</b></center></div><br>
<ul style="margin: 0 ; padding: 1 ;">
   <li>All data integrity verification are summarized in <a href="./data_integrity.csv"><div style="display:inline-block;color:blue;" "="">this csv file</div></a></li>
</ul><p>&nbsp;</p><h3><a name="iii2-importing-raw-data" class="md-header-anchor"></a><span>III.2. Importing raw data</span></h3><div style="text-align: justify"><span style="color:black">
   The step performs the import of sequencing data directly from DATAREF <b>into a QIIME 2 
   specific format</b>. In addition, <b>descriptive statistics of your data are generated
   </b>. The SAMBA workflow ran for you the commands below :
</span></div><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><span><span>​</span>x</span></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Import data using the q2_manifest ###</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools import \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> q2_manifest \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> data.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--type</span> <span class="cm-string">'SampleData[PairedEndSequencesWithQuality]'</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-format</span> PairedEndFastqManifestPhred33V2</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Descriptive statistics ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime demux summarize \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-data</span> data.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> data.qzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> data.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> import_output</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 460px;"></div><div class="CodeMirror-gutters" style="display: none; height: 460px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./import_output"><div style="display:inline-block;color:blue;" "="">import_output</div></a></center></b>
</div><p>&nbsp;</p><div style="background-color:yellow"><center><b>RESULTS</b></center></div><br>
<ul style="margin: 0 ; padding: 1 ;">
   <li>Sample repartition according to the sequence count</li>
</ul>  
<p align="center"> <img src="./import_output/demultiplex-summary.png" width="600" <="" p=""><br>
</p><ul style="margin: 0 ; padding: 1 ;">
   <li>All descriptive statistics of your samples are available <a href="./import_output/overview.html"><div style="display:inline-block;color:blue;" "="">here</div></a> (html output)</li>
</ul><p>&nbsp;</p><p>&nbsp;</p><h3><a name="iii3-removal-of-primers" class="md-header-anchor"></a><span>III.3. Removal of primers</span></h3><div style="text-align: justify"><span style="color:black">
   The third step is to remove the primers using <b>the cutadapt plugin</b> available in 
   QIIME 2. In addition, <b>descriptive statistics of your data are generated</b>. 
   The following commands made this step possible using the parameters that you have 
   defined in the parameters configuration file :
</span></div><br><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>primerF</li>
   <li>primerR</li>
   <li>errorRate</li>
   <li>overlap</li>
</ul><br><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Cutadapt process ###</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime cutadapt trim-paired \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-cores</span> <span class="cm-number">14</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-demultiplexed-sequences</span> data.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-front-f</span> <span class="cm-def">$primerF</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-front-r</span> <span class="cm-def">$primerR</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-error-rate</span> <span class="cm-def">$errorRate</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-discard-untrimmed</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-match-read-wildcards</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-overlap</span> <span class="cm-def">$overlap</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-trimmed-sequences</span> data_trimmed.qza</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Descriptive statistics ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize statistics of the cutadapt process</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime demux summarize \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-data</span> data_trimmed.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> data_trimmed.qzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> data_trimmed.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> trimmed_output &nbsp; </span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 598px;"></div><div class="CodeMirror-gutters" style="display: none; height: 598px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./trimmed_output"><div style="display:inline-block;color:blue;" "="">trimmed_output</div></a></center></b>
</div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br> 
<ul style="margin: 0 ; padding: 1 ;">
   <li>Sample repartition according to the sequence count</li></ul>
<p align="center"> <img src="./trimmed_output/demultiplex-summary.png" width="600" <="" p=""><br>
</p><ul style="margin:0 ; padding: 1 ;">
   <li>All descriptive statistics of your samples after the trimming are available <a href="./trimmed_output/overview.html"><div style="display:inline-block;color:blue;">here</div></a> (html output)</li>
   <li>The quality of your data before any quality filtering step is viewable <a href="./trimmed_output/quality-plot.html"><div style="display:inline-block;color:blue;">here</div></a> (html output)</li>
</ul>   <p>&nbsp;</p><h3><a name="iii4--sequence-quality-control-and-feature-table-construction" class="md-header-anchor"></a><span>III.4.  Sequence quality control and feature table construction</span></h3><div style="text-align: justify"><span style="color:black">
   The following step of the workflow is to filter the quality of the sequences according
   to the parameters defined yourself by having considered the quality of your data in the
   previous step. 
   Also to assemble the forward and reverse sequences, and to identify and remove the 
   chimeras. To performed this step, the <b>DADA2 R package</b> was used through QIIME 2 
   using the following commands :
</span></div><br><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>trim5F</li>
   <li>trim5R</li>
   <li>trunclenF</li>
   <li>trunclenR</li>
   <li>maxeeF</li>
   <li>maxeeR</li>
   <li>minqual</li>
   <li>chimeras</li>
</ul><p>&nbsp;</p><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### All DADA2 processes ###</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime dada2 denoise-paired \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-demultiplexed-seqs</span> data_trimmed.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-trim-left-f</span> <span class="cm-def">$trim5F</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-trim-left-r</span> <span class="cm-def">$trim5R</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-trunc-len-f</span> <span class="cm-def">$trunclenF</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-trunc-len-r</span> <span class="cm-def">$trunclenR</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-max-ee-f</span> <span class="cm-def">$maxeeF</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-max-ee-r</span> <span class="cm-def">$maxeeR</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-trunc-q</span> <span class="cm-def">$minqual</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-chimera-method</span> <span class="cm-def">$chimeras</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-n-threads</span> <span class="cm-number">14</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-representative-sequences</span> rep_seqs.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-table</span> table.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-denoising-stats</span> stats.qza</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Descriptive statistics ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize statistics of the DADA2 processes</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime metadata tabulate \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--m-input-file</span> stats.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> stats.qzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> stats.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dada2_output</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### ASV table ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Construction of the count table</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> table.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dada2_output &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize statistics from the table file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime feature-table summarize \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-table</span> table.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> table.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--m-sample-metadata-file</span> q2_metadata</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export of the table ASV statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> table.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dada2_output</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Reference sequences ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Obtaining ASV reference sequences</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime feature-table tabulate-seqs \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-data</span> rep_seqs.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> rep_seqs.qzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export of ASV reference sequences</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> rep_seqs.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dada2_output</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1403px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1403px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./dada2_output"><div style="display:inline-block;color:blue;" "="">dada2_output</div></a></center></b></div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>The dynamics of the different step of filtering can be visualized in this <a href="./dada2_output/index.html"><div style="display:inline-block;color:blue;">html file</div></a> and are also available in a <a href="./dada2_output/metadata.tsv"><div style="display:inline-block;color:blue;">tabulated file</div></a></li>
   <li>Distribution of sequences in samples</li>
        <p align="center"><img src="./dada2_output/sample-frequencies.png" width="600" <="" p=""><br>
   </p><li>Feature details are available <a href="./dada2_output/feature-frequency-detail.html"><div style="display:inline-block;color:blue;">here</div></a> (html output)</li>
   <li>Details about the samples can be found by going to <a href="./dada2_output/sample-frequency-detail.html"><div style="display:inline-block;color:blue;">this interactive html page</div></a></li>
           <li>Finally, you can retrieved the reference sequences of your ASVs in <a href="./dada2_output/sequences.fasta"><div style="display:inline-block;color:blue;">this fasta</div></a></li>
</ul><p>&nbsp;</p><p>&nbsp;</p><h3><a name="iii5--asv-clustering" class="md-header-anchor"></a><span>III.5.  ASV clustering</span></h3><div style="text-align: justify"><span style="color:black">
   The following step of the workflow is to cluster ASV according to sequence similarity and 
   abundance profil. This allow to takes into account the overestimation of diversity 
   produced by DADA2. It also reduces possible PCR errors. To performed this step, 
   the <b>dbOtu3 algorithm</b> was used through QIIME 2 using the following commands :
</span></div><br><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>genet_crit</li>
   <li>abund_crit</li>
   <li>pval_crit</li>
</ul><p>&nbsp;</p><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### dbOTU3 process ###</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime dbotu-q2 call-otus \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-table</span> table.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-sequences</span> rep_seqs.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-gen-crit</span> <span class="cm-def">$genet_crit</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-abund-crit</span> <span class="cm-def">$abund_crit</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-pval-crit</span> <span class="cm-def">$pval_crit</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-representative-sequences</span> dbotu3_seqsqza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-dbotu-table</span> dbotu3_tableqza &gt; dbotu3_details</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### ASV table ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Construction of the count table</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> dbotu3_tableqza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dbotu3_output &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize statistics from the table file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime feature-table summarize \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-table</span> dbotu3_tableqza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> dbotu3_tableqzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--m-sample-metadata-file</span> q2_metadata</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export of the table ASV statistics</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> dbotu3_tableqzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dbotu3_output</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Reference sequences ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Obtaining ASV reference sequences</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime feature-table tabulate-seqs \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--verbose</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-data</span> dbotu3_seqsqza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> dbotu3_seqsqzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export of ASV reference sequences</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> dbotu3_seqsqzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> dbotu3_output</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 966px;"></div><div class="CodeMirror-gutters" style="display: none; height: 966px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./dbotu3_output"><div style="display:inline-block;color:blue;" "="">dbotu3_output</div></a></center></b></div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>Overview of the results available <a href="./dbotu3_output/index.html"><div style="display:inline-block;color:blue;">here</div></a> (html output)</li>
   <li>Feature details are available <a href="./dbotu3_output/feature-frequency-detail.html"><div style="display:inline-block;color:blue;">here</div></a> (html output)</li>
   <li>Details about the samples can be found by going to <a href="./dbotu3_output/sample-frequency-detail.html"><div style="display:inline-block;color:blue;">this interactive html page</div></a></li>
           <li>Finally, you can retrieved the reference sequences of your ASVs in <a href="./dbotu3_output/sequences.fasta"><div style="display:inline-block;color:blue;">this fasta</div></a></li>
</ul><p>&nbsp;</p><p>&nbsp;</p><h3><a name="iii6--taxonomic-assignation" class="md-header-anchor"></a><span>III.6.  Taxonomic assignation</span></h3><div style="text-align: justify">
   The taxonomic assignation performed during the fourth step of the workflow allowed to 
   affiliate each ASV to a taxonomy by using as reference the database defined by yourself
</div><br><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>database</li>
   <li>confidence</li>
</ul><p>&nbsp;</p><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Taxonomic assignation ###</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime feature-classifier classify-sklearn \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-n-jobs</span> <span class="cm-number">14</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--p-confidence</span> <span class="cm-def">$confidence</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-classifier</span> <span class="cm-def">$database</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--i-reads</span> rep_seqs.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-classification</span> taxonomy.qza</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">### Output formatting ###</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Summarize taxonomic results</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime metadata tabulate \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--m-input-file</span> taxonomy.qza \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--o-visualization</span> taxonomy.qzv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Export</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">qiime tools <span class="cm-keyword">export</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--input-path</span> taxonomy.qzv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--output-path</span> taxo_output</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Formatting</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">mv</span> taxo_output/metadata.tsv ASV_taxonomy.tsv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sed</span> <span class="cm-attribute">-i</span> <span class="cm-string">'1,2d'</span> ASV_taxonomy.tsv</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sed</span> <span class="cm-attribute">-i</span> <span class="cm-string">'1 i\#OTUID\ttaxonomy\tconfidence'</span> ASV_taxonomy.tsv</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 575px;"></div><div class="CodeMirror-gutters" style="display: none; height: 575px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./taxo_output"><div style="display:inline-block;color:blue;" "="">taxo_output</div></a></center></b></div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>The result of the taxonomic affiliation is available in <a href="./taxo_output/index.html"><div style="display:inline-block;color:blue;">html</div></a> and <a href="./taxo_output/ASV_taxonomy.tsv"><div style="display:inline-block;color:blue;">tabulated</div></a> formats</li>
</ul> <p>&nbsp;</p><h3><a name="iii7--samples-decontamination-microdecon-optional" class="md-header-anchor"></a><span>III.7.  Samples decontamination (microDecon) [optional]</span></h3><div style="text-align: justify">
If you have sequenced control samples, this step allows you to decontaminate your samples from the sequences identified in these controls. This step is carried out using microDecon and a homemade R script
</div><br><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>control_list</li>
   <li>nb_controls</li>
   <li>nb_samples</li>   
</ul><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./microDecon"><div style="display:inline-block;color:blue;" "="">microDecon</div></a></center></b></div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>New ASV table: <a href="./microDecon/decontaminated_ASV_table.tsv"><div style="display:inline-block;color:blue;">decontaminated table</div></a></li>
   <li>Representative sequence: <a href="./microDecon/decontaminated_ASV.fasta"><div style="display:inline-block;color:blue;">decontaminated ASV sequence</div></a></li>
   <li>All ASVs listed <a href="./microDecon/ASV_removed.txt"><div style="display:inline-block;color:blue;">here</div></a> were removed because they were in the majority in the control samples</li>
    <li>For all ASVs listed <a href="./microDecon/abundance_removed.txt"><div style="display:inline-block;color:blue;">here</div></a>, a proportion of sequences were removed in each sample. Note that this file also contains the ASVs unique to the controls which have therefore been removed.</li>
</ul> <p>&nbsp;</p><h3><a name="iii8--functional-predictions-picrust2-optional" class="md-header-anchor"></a><span>III.8.  Functional predictions (PICRUSt2) [optional]</span></h3><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS USED</span></b>
   <li>method</li>
   <li>nsti</li>   
</ul><p>&nbsp;</p><div style="background-color:rgba(135, 206, 250, 0.6)"><b><center>OUTPUT FOLDER: <a href="./picrust2_output"><div style="display:inline-block;color:blue;" "="">picrust2_output</div></a></center></b></div><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
    <li>EC predictions: <a href="./picrust2_output/q2-picrust2_output/ec_metagenome_exported/ec_metagenome_predictions_with-descriptions.tsv"><div style="display:inline-block;color:blue;">EC metagenome predictions</div></a></li>
    <li>KO predictions: <a href="./picrust2_output/q2-picrust2_output/ko_metagenome_exported/ko_metagenome_predictions_with-descriptions.tsv"><div style="display:inline-block;color:blue;">KO metagenome predictions</div></a></li>
    <li>METACYC predictions: <a href="./picrust2_output/q2-picrust2_output/pathway_abundance_exported/pathway_abundance_predictions_with-descriptions.tsv"><div style="display:inline-block;color:blue;">METACYC abundance predictions</div></a></li></ul><p align="center"> <img src="./picrust2_output/EC_functional_predictions_NMDS.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./picrust2_output/KO_functional_predictions_NMDS.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./picrust2_output/MetaCyc_functional_predictions_NMDS.png" width="1000" <="" p=""></p><p>&nbsp;</p><h3><a name="iii9--final-outputs" class="md-header-anchor"></a><span>III.9.  Final outputs</span></h3><div style="text-align: justify"><span style="color:black">
   This is the last step of the bioinformatic process where the goal is to merge the ASV abundance table with the taxonomy file. This is performed using the following commands :
</span></div><pre spellcheck="false" class="md-fences md-end-block ty-contain-cm modeLoaded" lang="bash"><div class="CodeMirror cm-s-inner CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 8px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">biom add-metadata \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">-i</span> dada2_output/feature-table.biom \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--observation-metadata-fp</span> ASV_taxonomy.tsv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">-o</span> Final_ASV_table_with_taxonomy.biom \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--sc-separated</span> taxonomy</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">biom convert \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">-i</span> Final_ASV_table_with_taxonomy.biom \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">-o</span> Final_ASV_table_with_taxonomy.tsv \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--to-tsv</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-attribute">--header-key</span> taxonomy</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 253px;"></div><div class="CodeMirror-gutters" style="display: none; height: 253px;"></div></div></div></pre><p>&nbsp;</p><div style="background-color:yellow" ;=""><b><center>RESULTS</center></b></div><br>
<ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>The final ASV table can be viewed <a href="./ASV_table_with_taxonomy.tsv"><div style="display:inline-block;color:blue;">here</div></a>. A biom file is available <a href="./ASV_table_with_taxonomy.biom"><div style="display:inline-block;color:blue;">here</div></a> formats</li>
</ul> <p>&nbsp;</p><h2><a name="span-stylecolorrediv-general-statistical-analysesspan" class="md-header-anchor"></a><span style="color:red"><span>IV. General statistical analyses</span></span></h2><ul style="border: 2px dashed #FF0000 ;margin:0;padding:1;"><b><span style="color:red;">CONFIGURATION SETTINGS</span></b>
   <li>Normalization method : "None / Rarefaction / DESeq2 / CSS"</li>
   <li>Grouping variable</li>
   <li>Distance matrix : "Jaccard / Bray-Curtis / UniFrac / Weighted UniFrac"</li>
</ul><p>&nbsp;</p><h3><a name="iv1--alpha-diversity" class="md-header-anchor"></a><span>IV.1.  Alpha diversity</span></h3><h4><a name="iiv1a-diversity-indicesi" class="md-header-anchor"></a><i><span>IV.1.a diversity indices</span></i></h4><p align="center"> <img src="./R/FIGURES/alpha_diversity/diversity_index/alpha_div_plots.png" width="1000" <="" p=""><br></p><h4><a name="iiv1b-rarefaction-curvei" class="md-header-anchor"></a><i><span>IV.1.b rarefaction curve</span></i></h4><p align="center"> <img src="./R/FIGURES/alpha_diversity/rarefaction_curve.png" width="1000" <="" p=""><br></p><h4><a name="iiv1c-taxonomic-diversityi" class="md-header-anchor"></a><i><span>IV.1.c taxonomic diversity</span></i></h4><p align="center"> <img src="./R/FIGURES/alpha_diversity/diversity_barplots/barplot_phylum.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/alpha_diversity/diversity_barplots/barplot_class.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/alpha_diversity/diversity_barplots/barplot_order.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/alpha_diversity/diversity_barplots/barplot_genus.png" width="1000" <="" p=""><br>
</p><ul style="margin: 0 ; padding: 1 ;text-align: justify">
   <li>The results of the significance tests carried out for each variable on each diversity indexes can be viewed <a href="./R/FIGURES/alpha_diversity/index_significance_tests.txt"><div style="display:inline-block;color:blue;">here</div></a></li>
</ul> <p>&nbsp;</p><h3><a name="iv2--beta-diversity" class="md-header-anchor"></a><span>IV.2.  Beta diversity</span></h3><div style="text-align: justify"><span style="color:black">
   <font color="red"><b>NOTE 1: Raw results without any data normalization are available 
      <a href="./R/FIGURES/beta_diversity_non_normalized"><div style="display:inline-block;color:blue;">here</div></a>. However, we do not recommend using them because from a statistical point of view, these results are very questionable.</b>
   </font><br><br>
   <font color="red"><b>NOTE 2: For the beta diversity, SAMBA calculates 4 different distance 
      matrices with jaccard, bray-curtis, UniFrac and weighted UniFrac indices. To facilitate 
      the reading of this report, we have displayed that the results of the bray-curtis and 
      weighted UniFrac matrices. All results are available in each normalization folder <a href="./R/FIGURES/"><div style="display:inline-block;color:blue;">here</div></a></b>
   </font>
</span></div><p>&nbsp;</p><h4><a name="iiv2a-nmdsi" class="md-header-anchor"></a><i><span>IV.2.a NMDS</span></i><span> </span></h4><h6><a name="iv2ai-rarefied-data" class="md-header-anchor"></a><span>IV.2.a.i rarefied data</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/NMDS/NMDS_rarefied_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/NMDS/NMDS_rarefied_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2aii-deseq2-normalization" class="md-header-anchor"></a><span>IV.2.a.ii DESeq2 normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/NMDS/NMDS_DESeq2_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/NMDS/NMDS_DESeq2_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2aiii-css-normalization" class="md-header-anchor"></a><span>IV.2.a.iii CSS normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/NMDS/NMDS_CSS_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/NMDS/NMDS_CSS_wunifrac.png" width="1000" <="" p=""><br></p><h4><a name="iiv2b-pcoai" class="md-header-anchor"></a><i><span>IV.2.b PCoA</span></i><span> </span></h4><h6><a name="iv2bi-rarefied-data" class="md-header-anchor"></a><span>IV.2.b.i rarefied data</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/PCoA/PCoA_rarefied_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/PCoA/PCoA_rarefied_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2bii-deseq2-normalization" class="md-header-anchor"></a><span>IV.2.b.ii DESeq2 normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/PCoA/PCoA_DESeq2_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/PCoA/PCoA_DESeq2_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2biii-css-normalization" class="md-header-anchor"></a><span>IV.2.b.iii CSS normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/PCoA/PCoA_CSS_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/PCoA/PCoA_CSS_wunifrac.png" width="1000" <="" p=""><br></p><h4><a name="iiv2c-hierarchical-clusteringi" class="md-header-anchor"></a><i><span>IV.2.c Hierarchical clustering</span></i><span> </span></h4><h6><a name="iv2ci-rarefied-data" class="md-header-anchor"></a><span>IV.2.c.i rarefied data</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/Hierarchical_Clustering/hclustering_rarefied_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/Hierarchical_Clustering/hclustering_rarefied_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2cii-deseq2-normalization" class="md-header-anchor"></a><span>IV.2.c.ii DESeq2 normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/Hierarchical_Clustering/hclustering_DESeq2_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/Hierarchical_Clustering/hclustering_DESeq2_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2ciii-css-normalization" class="md-header-anchor"></a><span>IV.2.c.iii CSS normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/Hierarchical_Clustering/hclustering_CSS_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/Hierarchical_Clustering/hclustering_CSS_wunifrac.png" width="1000" <="" p=""><br></p><h4><a name="iiv2d-explained-variancei" class="md-header-anchor"></a><i><span>IV.2.d Explained variance</span></i><span> </span></h4><h6><a name="iv2di-rarefied-data" class="md-header-anchor"></a><span>IV.2.d.i rarefied data</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/pie_ExpVar_rarefied_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_rarefied/pie_ExpVar_rarefied_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2dii-deseq2-normalization" class="md-header-anchor"></a><span>IV.2.d.ii DESeq2 normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/pie_ExpVar_DESeq2_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_DESeq2/pie_ExpVar_DESeq2_wunifrac.png" width="1000" <="" p=""><br></p><h6><a name="iv2diii-css-normalization" class="md-header-anchor"></a><span>IV.2.d.iii CSS normalization</span></h6><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/pie_ExpVar_CSS_bray.png" width="1000" <="" p=""><br>
</p><p align="center"> <img src="./R/FIGURES/beta_diversity_CSS/pie_ExpVar_CSS_wunifrac.png" width="1000" <="" p=""><br></p></div>
</body>
</html>